### PR TITLE
fix diagnostic feature and code cleanup

### PR DESCRIPTION
**Pull request by librasungirl**
Update network diagnostic to truly pass 5.7.1 Router 
[Router_5_7_1.zip](https://github.com/openthread/openthread/files/624187/Router_5_7_1.zip)

@jwhui please help to review it.

### CODE REVIEW

File: src/cli/cli.cpp [hunk 1]
```diff
@@ -2626,29 +2626,34 @@ void Interpreter::ProcessNetworkDiagnostic(int argc, char *argv[])
 {
     ThreadError error = kThreadError_None;
     struct otIp6Address address;
-    uint8_t index = 2;
-    uint8_t tlvTypes[OT_NUM_NETDIAG_TLV_TYPES];
-    uint8_t count = 0;
+    uint8_t payload[2 + OT_NETWORK_DIAGNOSTIC_TYPELIST_MAX_ENTRIES];  // TypeList Type(1B), len(1B), type list
+    uint8_t payloadIndex = 0;
+    uint8_t paraIndex = 0;
```
Comment (jwhui) [inline src/cli/cli.cpp]: nit: rename to something like `paramIndex`?  Took me a while to realize this was short for "parameter index".

File: src/core/mac/mac.hpp [hunk 2]
```diff
@@ -493,6 +494,22 @@ class Mac
     void SetPromiscuous(bool aPromiscuous);
 
     /**
+     * This function fills network diagnostic MacCounterTlv.
+     *
+     *
+     * @param[in]  aMacCountersTlv The reference to the network diagnostic MacCounterTlv.
+     *
+     */
+    void FillMacCountersTlv(NetworkDiagnostic::MacCountersTlv &aMacCounters);
```
Comment (jwhui) [inline src/core/mac/mac.hpp]: Declare as `const`?

File: src/core/mac/mac.hpp [hunk 3]
```diff
@@ -493,6 +494,22 @@ class Mac
     void SetPromiscuous(bool aPromiscuous);
 
     /**
+     * This function fills network diagnostic MacCounterTlv.
+     *
+     *
```
Comment (jwhui) [inline src/core/mac/mac.hpp]: nit: extra line

File: src/core/mac/mac.hpp [hunk 4]
```diff
@@ -493,6 +494,22 @@ class Mac
...
+     * @param[in]  aMacCountersTlv The reference to the network diagnostic MacCounterTlv.
+     *
+     */
+    void FillMacCountersTlv(NetworkDiagnostic::MacCountersTlv &aMacCounters);
+
+    /**
+     * This function resets mac counters
+     *
+     *
```
Comment (jwhui) [inline src/core/mac/mac.hpp]: nit: extra line

**Comment by bukepo** [comment]
LGTM

**Pull request merged**